simpleast-core/src/main/java/com/discord/simpleast/code/JavaScript.kt

```diff
@@ -0,0 +1,168 @@
+    pre: String, signature: String?, params: String,
+    codeStyleProviders: CodeStyleProviders<RC>
+  ) : Node.Parent<RC>(
+      StyleNode.TextStyledNode(pre, codeStyleProviders.keywordStyleProvider),
+      signature?.let { StyleNode.TextStyledNode(signature, codeStyleProviders.identifierStyleProvider) },
+      StyleNode.TextStyledNode(params, codeStyleProviders.paramsStyleProvider)
+  ) {
+      companion object {
+        /**
+         * Matches against a JavaScript function declaration.
+         *
+         * ```
+         * function foo(bar)
+         * function baz()
+         * ```
+         */
+         private val PATTERN_JAVASCRIPT_FUNC = 
+             """^(function\*?|static|get|set|async)(\s+\w+)?( *?\(.*?\))""".toRegex(RegexOption.DOT_MATCHES_ALL).toPattern()
+
+         fun <RC, S> createFunctionRule(codeStyleProviders: CodeStyleProviders<RC>) =
+          object : Rule<RC, Node<RC>, S>(PATTERN_JAVASCRIPT_FUNC) {
+            override fun parse(matcher: Matcher, parser: Parser<RC, in Node<RC>, S>, state: S): ParseSpec<RC, S> {
+              val definition = matcher.group(1)
+              val signature = matcher.group(2)
+              val params = matcher.group(3)
+              return ParseSpec.createTerminal(FunctionNode(definition!!, signature, params!!, codeStyleProviders), state)
+            }
+          }
+    }
+  }
+
+  class FieldNode<RC>(
+      definition: String, name: String,
+      codeStyleProviders: CodeStyleProviders<RC>
+  ) : Node.Parent<RC>(
+      StyleNode.TextStyledNode(definition, codeStyleProviders.keywordStyleProvider),
+      StyleNode.TextStyledNode(name, codeStyleProviders.identifierStyleProvider),
+  ) {
+    companion object {
+      /**
+       * Matches against a JavaScript field definition.
+       *
+       * ```
+       * var x = 1;
+       * let y = 5;
+       * const z = 10;
+       * ```
+       */
+      private val PATTERN_JAVASCRIPT_FIELD =
+          Pattern.compile("""^(var|let|const)(\s+\w+)""", Pattern.DOTALL)
+
+      fun <RC, S> createFieldRule(
+          codeStyleProviders: CodeStyleProviders<RC>
+      ) =
+          object : Rule<RC, Node<RC>, S>(PATTERN_JAVASCRIPT_FIELD) {
+            override fun parse(matcher: Matcher, parser: Parser<RC, in Node<RC>, S>, state: S):
+                ParseSpec<RC, S> {
+              val definition = matcher.group(1)
+              val name = matcher.group(2)
+              return ParseSpec.createTerminal(
+                  FieldNode(definition!!, name!!, codeStyleProviders), state)
+            }
+          }
+    }
+  }
+
+  /**
+   * Matches against a JavaScript regex.
+   *
+   * ```
+   * /(.*)/
+   * ```
+   */
+   private val PATTERN_JAVASCRIPT_REGEX = 
+       Pattern.compile("""^/.*?/(?:\w*)?""", Pattern.DOTALL)
+
+  /**
+   * Matches against a JavaScript generic.
+   *
+   * ```
+   * <pending>
+   * ```
+   */
+  private val PATTERN_JAVASCRIPT_GENERIC = 
+      Pattern.compile("""^<(.*)>""", Pattern.DOTALL)
```

Comment:
```suggestion
      Pattern.compile("""^<(.*)>""")
```
Don't want this matching line terminators